ocamlPackages.janestreet: 0.15 -> 0.16

dimitrijer opened the pull request:
Add version 0.16 of Jane Street OCaml libraries. Note that this does not replace the existing 0.15 set of libraries.

Notable omissions:

- [hardcaml](https://github.com/janestreet/hardcaml) libraries weren't present in 0.15, I plan to add them in a separate PR.
- [torch ](https://github.com/janestreet/torch) library is already provided under `ocamlPackages.torch`. Recently Jane Street took over development of said library. At the moment, both old and new version of the library are still in use, because the new version does not support PyTorch 2.0.0. The new, JS-maitained torch library should be added under `ocamlPackages.torch`, and is out of scope of this PR.

Other notes:
- pythonlib is marked as broken due to https://github.com/janestreet/pythonlib/issues/18
- several patches that were necessary to build some 0.15 libraries (e.g. bonsai and incr_dom) are no longer necessary for 0.16

## Description of changes

Release announcement: https://discuss.ocaml.org/t/ann-v0-16-release-of-jane-street-packages/12398

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

vbgl commented:
This seems to break a few things (e.g., `comby`) and a few hashes might be wrong (e.g., `streamable`). Also, what’s the relation with #230267?

dimitrijer commented:
Sorry about that, you're right: more than a few packages are broken. I ran the following:

```
$ nix-shell -p nixpkgs-review --run "nixpkgs-review pr 247022"
<snip>
11 packages failed to build:
cmdstan comby coqPackages.coq-lsp coqPackages.serapi libbap ligo python310Packages.bap python310Packages.bap.dist python311Packages.bap python311Packages.bap.dist stanc

41 packages built:
acgtk advi alt-ergo beluga coccinelle coqPackages.coq-elpi coqPackages.hierarchy-builder coqPackages.mathcomp-algebra-tactics coqPackages.mathcomp-analysis coqPackages.mathcomp-classical coqPackages.mathcomp-infotheo coqPackages.trakt cvehound cvehound.dist dune-release easycrypt eff flitter flow framac fstar glsurf jackline liquidsoap ocamlfor
mat ocamlformat_0_19_0 ocamlformat_0_20_0 ocamlformat_0_20_1 ocamlformat_0_21_0 ocamlformat_0_22_4 ocamlformat_0_23_0 ocamlformat_0_24_1 ocamlformat_0_25_1 patdiff reason sail-riscv-rv32 sail-riscv-rv64 satysfi soupault wayland-proxy-virtwl why3
```

I am slowly going through failures and patching where viable. I think `bap` and dependencies are also failing in `master` as well, so I'll ignore those. I also think patching some of them (`stanc`, `ligo`) is a big effort and should be done upstream, in which case I'll try to provide previous (0.15) versions of libraries to their derivations.

I'm quite new to Nix, sorry about the failures again! Hate to waste reviewiers time.

Regarding #230267, I was made aware of this PR after I submitted this one. Seems like it's been inactive for a while. If there's a way to retrofit that PR with changes in this one, I'm happy to do so.

Alizter commented:
Result of `nixpkgs-review pr 247022` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages failed to build:</summary>
  <ul>
    <li>libbap</li>
    <li>python310Packages.bap</li>
    <li>python310Packages.bap.dist</li>
    <li>python311Packages.bap</li>
    <li>python311Packages.bap.dist</li>
  </ul>
</details>
<details>
  <summary>44 packages built:</summary>
  <ul>
    <li>acgtk</li>
    <li>advi</li>
    <li>alt-ergo</li>
    <li>beluga</li>
    <li>coccinelle</li>
    <li>comby</li>
    <li>coqPackages.coq-elpi</li>
    <li>coqPackages.coq-lsp</li>
    <li>coqPackages.hierarchy-builder</li>
    <li>coqPackages.mathcomp-algebra-tactics</li>
    <li>coqPackages.mathcomp-analysis</li>
    <li>coqPackages.mathcomp-classical</li>
    <li>coqPackages.mathcomp-infotheo</li>
    <li>coqPackages.serapi</li>
    <li>coqPackages.trakt</li>
    <li>cvehound</li>
    <li>cvehound.dist</li>
    <li>dune-release</li>
    <li>easycrypt</li>
    <li>eff</li>
    <li>flitter</li>
    <li>flow</li>
    <li>framac</li>
    <li>fstar</li>
    <li>glsurf</li>
    <li>jackline</li>
    <li>liquidsoap</li>
    <li>ocamlformat (ocamlformat_0_26_0)</li>
    <li>ocamlformat_0_19_0</li>
    <li>ocamlformat_0_20_0</li>
    <li>ocamlformat_0_20_1</li>
    <li>ocamlformat_0_21_0</li>
    <li>ocamlformat_0_22_4</li>
    <li>ocamlformat_0_23_0</li>
    <li>ocamlformat_0_24_1</li>
    <li>ocamlformat_0_25_1</li>
    <li>patdiff</li>
    <li>reason</li>
    <li>sail-riscv-rv32</li>
    <li>sail-riscv-rv64</li>
    <li>satysfi</li>
    <li>soupault</li>
    <li>wayland-proxy-virtwl</li>
    <li>why3</li>
  </ul>
</details>

dimitrijer commented:
All packages except `bap` and its dependencies build now.

I fixed the hash for `streamable`, apparently a new commit was pushed to v0.16 branch in the meantime.

I also tested the build on x86_64-darwin.

Could I get some help in making `ocamlPackages.bap` use attributes defined in `ocamlPackages.janeStreet_0_15`? I have the following in `pkgs/top-level/ocaml-packages.nix`:

```nix
    bap = callPackage ../development/ocaml-modules/bap {
      inherit (pkgs.llvmPackages) llvm;
      inherit (janeStreet_0_15) core_kernel ppx_jane;
    };
```

I believe the issue is that libraries in `ocamlPackages.janeStreet_0_15` pick up build dependencies from outer scope, which contains 0.16 libraries. For example, running `nix-build ./ -A ocamlPackages.janeStreet_0_15.core_kernel` fails because it pulls in 0.16 libraries. How would I go about fixing this?

dimitrijer commented:
Everything except broken packages should build now. Here's the latest build report on x86_64-darwin:

```
$ nix-shell -p nixpkgs-review --run "nixpkgs-review pr 247022"
<snip>
3 packages marked as broken and skipped:
flitter glsurf liquidsoap

43 packages built:
acgtk advi alt-ergo beluga coccinelle comby coqPackages.coq-elpi coqPackages.coq-lsp coqPackages.hie
rarchy-builder coqPackages.mathcomp-algebra-tactics coqPackages.mathcomp-analysis coqPackages.mathco
mp-classical coqPackages.mathcomp-infotheo coqPackages.serapi coqPackages.trakt cvehound cvehound.di
st dune-release easycrypt eff flow framac fstar jackline libbap ocamlformat ocamlformat_0_19_0 ocaml
format_0_20_0 ocamlformat_0_20_1 ocamlformat_0_21_0 ocamlformat_0_22_4 ocamlformat_0_23_0 ocamlforma
t_0_24_1 ocamlformat_0_25_1 patdiff python310Packages.bap python310Packages.bap.dist python311Packag
es.bap python311Packages.bap.dist reason satysfi soupault why3
```

vbgl commented:
Awesome! The hash of `ppx_accessor` seems wrong.

vbgl commented:
Here is a patch to fix `torch`:

~~~
--- a/pkgs/development/ocaml-modules/torch/default.nix
+++ b/pkgs/development/ocaml-modules/torch/default.nix
@@ -57,7 +57,6 @@ buildDunePackage rec {
   preBuild = "export LIBTORCH=${torch.dev}/";

   doCheck = !stdenv.isAarch64;
-  checkPhase = "dune runtest";

   meta = with lib; {
     inherit (src.meta) homepage;
~~~

dimitrijer commented:
Fixed hash for `ppx_accessor`, thanks!

I also moved `ppx_bap`, `bap` and `torch` under `janeStreet_0_15` scope. That way `callPackage` will pick up correct dependencies from the scope automatically. It's much more concise now.

dimitrijer commented:
Latest build report on x86_64-linux:

```
51 packages built:
acgtk advi alt-ergo beluga coccinelle comby coqPackages.coq-elpi coqPackages.coq-lsp coqPackages.hierarchy-builder coqPackages.mathcomp-algebra-tactics coqPackages.mathcomp-analysis coqPackages.mathcomp-classical coqPackages.mathcomp-infotheo coqPackages.serapi coqPackages.trakt cvehound cvehound.dist dune-release easycrypt eff flitter flow framac fstar glsurf hol_light jackline libbap liquidsoap merge-fmt ocamlformat ocamlformat_0_19_0 ocamlformat_0_20_0 ocamlformat_0_20_1 ocamlformat_0_21_0 ocamlformat_0_22_4 ocamlformat_0_23_0 ocamlformat_0_24_1 ocamlformat_0_25_1 patdiff python310Packages.bap python310Packages.bap.dist python311Packages.bap python311Packages.bap.dist reason sail-riscv-rv32 sail-riscv-rv64 satysfi soupault wayland-proxy-virtwl why3
```

@vbgl  torch seems to build fine on tip, I left the tests enabled.

Alizter commented:
Result of `nixpkgs-review pr 247022` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>51 packages built:</summary>
  <ul>
    <li>acgtk</li>
    <li>advi</li>
    <li>alt-ergo</li>
    <li>beluga</li>
    <li>coccinelle</li>
    <li>comby</li>
    <li>coqPackages.coq-elpi</li>
    <li>coqPackages.coq-lsp</li>
    <li>coqPackages.hierarchy-builder</li>
    <li>coqPackages.mathcomp-algebra-tactics</li>
    <li>coqPackages.mathcomp-analysis</li>
    <li>coqPackages.mathcomp-classical</li>
    <li>coqPackages.mathcomp-infotheo</li>
    <li>coqPackages.serapi</li>
    <li>coqPackages.trakt</li>
    <li>cvehound</li>
    <li>cvehound.dist</li>
    <li>dune-release</li>
    <li>easycrypt</li>
    <li>eff</li>
    <li>flitter</li>
    <li>flow</li>
    <li>framac</li>
    <li>fstar</li>
    <li>glsurf</li>
    <li>hol_light</li>
    <li>jackline</li>
    <li>libbap</li>
    <li>liquidsoap</li>
    <li>merge-fmt</li>
    <li>ocamlformat (ocamlformat_0_26_0)</li>
    <li>ocamlformat_0_19_0</li>
    <li>ocamlformat_0_20_0</li>
    <li>ocamlformat_0_20_1</li>
    <li>ocamlformat_0_21_0</li>
    <li>ocamlformat_0_22_4</li>
    <li>ocamlformat_0_23_0</li>
    <li>ocamlformat_0_24_1</li>
    <li>ocamlformat_0_25_1</li>
    <li>patdiff</li>
    <li>python310Packages.bap</li>
    <li>python310Packages.bap.dist</li>
    <li>python311Packages.bap</li>
    <li>python311Packages.bap.dist</li>
    <li>reason</li>
    <li>sail-riscv-rv32</li>
    <li>sail-riscv-rv64</li>
    <li>satysfi</li>
    <li>soupault</li>
    <li>wayland-proxy-virtwl</li>
    <li>why3</li>
  </ul>
</details>

georgyo commented:
Tested a few of these packages manually to ensure things are working. Things look pretty good to me.

ulrikstrid reviewed (approved):
Seems ok, since @vbgl has been reviewing previously I'll let him merge

vbgl commented on pkgs/top-level/ocaml-packages.nix:
```diff
@@ -747,6 +752,40 @@ let
+
+      ppx_bap = callPackage ../development/ocaml-modules/ppx_bap { };
+
+      torch = callPackage ../development/ocaml-modules/torch {
+        inherit (pkgs.python3Packages) torch;
+      };
```
Please leave `torch` alone (see #249474).

dimitrijer replied:
Ack, reverted changes to torch.

Alizter commented:
Result of `nixpkgs-review pr 247022` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>ocamlPackages.pythonlib</li>
  </ul>
</details>
<details>
  <summary>6 packages failed to build:</summary>
  <ul>
    <li>ocamlPackages.biocaml</li>
    <li>ocamlPackages.bistro</li>
    <li>ocamlPackages.cfstream</li>
    <li>ocamlPackages.magic-trace</li>
    <li>ocamlPackages.phylogenetics</li>
    <li>ocamlPackages.tls-async</li>
  </ul>
</details>
<details>
  <summary>527 packages built:</summary>
  <ul>
    <li>acgtk</li>
    <li>advi</li>
    <li>alt-ergo</li>
    <li>beluga</li>
    <li>coccinelle</li>
    <li>comby</li>
    <li>coqPackages.coq-elpi</li>
    <li>coqPackages.coq-lsp</li>
    <li>coqPackages.hierarchy-builder</li>
    <li>coqPackages.mathcomp-algebra-tactics</li>
    <li>coqPackages.mathcomp-analysis</li>
    <li>coqPackages.mathcomp-classical</li>
    <li>coqPackages.mathcomp-infotheo</li>
    <li>coqPackages.serapi</li>
    <li>coqPackages.trakt</li>
    <li>cvehound</li>
    <li>cvehound.dist</li>
    <li>dune-release</li>
    <li>easycrypt</li>
    <li>eff</li>
    <li>flitter</li>
    <li>flow</li>
    <li>framac</li>
    <li>fstar</li>
    <li>glsurf</li>
    <li>hol_light</li>
    <li>jackline</li>
    <li>libbap</li>
    <li>liquidsoap</li>
    <li>merge-fmt</li>
    <li>ocamlPackages.abstract_algebra</li>
    <li>ocamlPackages.accessor</li>
    <li>ocamlPackages.accessor_async</li>
    <li>ocamlPackages.accessor_base</li>
    <li>ocamlPackages.accessor_core</li>
    <li>ocamlPackages.alcotest-lwt</li>
    <li>ocamlPackages.alcotest-mirage</li>
    <li>ocamlPackages.angstrom</li>
    <li>ocamlPackages.angstrom-async</li>
    <li>ocamlPackages.angstrom-lwt-unix</li>
    <li>ocamlPackages.angstrom-unix</li>
    <li>ocamlPackages.arp</li>
    <li>ocamlPackages.async</li>
    <li>ocamlPackages.async_durable</li>
    <li>ocamlPackages.async_extra</li>
    <li>ocamlPackages.async_find</li>
    <li>ocamlPackages.async_inotify</li>
    <li>ocamlPackages.async_interactive</li>
    <li>ocamlPackages.async_js</li>
    <li>ocamlPackages.async_kernel</li>
    <li>ocamlPackages.async_rpc_kernel</li>
    <li>ocamlPackages.async_rpc_websocket</li>
    <li>ocamlPackages.async_sendfile</li>
    <li>ocamlPackages.async_shell</li>
    <li>ocamlPackages.async_smtp</li>
    <li>ocamlPackages.async_ssl</li>
    <li>ocamlPackages.async_unix</li>
    <li>ocamlPackages.async_websocket</li>
    <li>ocamlPackages.awa</li>
    <li>ocamlPackages.awa-mirage</li>
    <li>ocamlPackages.babel</li>
    <li>ocamlPackages.bap</li>
    <li>ocamlPackages.base</li>
    <li>ocamlPackages.base64</li>
    <li>ocamlPackages.base_bigstring</li>
    <li>ocamlPackages.base_quickcheck</li>
    <li>ocamlPackages.base_trie</li>
    <li>ocamlPackages.bidirectional_map</li>
    <li>ocamlPackages.bignum</li>
    <li>ocamlPackages.bin_prot</li>
    <li>ocamlPackages.bisect_ppx</li>
    <li>ocamlPackages.bls12-381</li>
    <li>ocamlPackages.bls12-381-signature</li>
    <li>ocamlPackages.bonsai</li>
    <li>ocamlPackages.bos</li>
    <li>ocamlPackages.brisk-reconciler</li>
    <li>ocamlPackages.brr</li>
    <li>ocamlPackages.ca-certs</li>
    <li>ocamlPackages.ca-certs-nss</li>
    <li>ocamlPackages.camlimages</li>
    <li>ocamlPackages.camlp5</li>
    <li>ocamlPackages.caqti</li>
    <li>ocamlPackages.caqti-async</li>
    <li>ocamlPackages.caqti-driver-mariadb</li>
    <li>ocamlPackages.caqti-driver-postgresql</li>
    <li>ocamlPackages.caqti-driver-sqlite3</li>
    <li>ocamlPackages.caqti-dynload</li>
    <li>ocamlPackages.caqti-lwt</li>
    <li>ocamlPackages.caqti-type-calendar</li>
    <li>ocamlPackages.carton</li>
    <li>ocamlPackages.carton-git</li>
    <li>ocamlPackages.carton-lwt</li>
    <li>ocamlPackages.checkseum</li>
    <li>ocamlPackages.cinaps</li>
    <li>ocamlPackages.class_group_vdf</li>
    <li>ocamlPackages.cohttp</li>
    <li>ocamlPackages.cohttp-async</li>
    <li>ocamlPackages.cohttp-lwt</li>
    <li>ocamlPackages.cohttp-lwt-jsoo</li>
    <li>ocamlPackages.cohttp-lwt-unix</li>
    <li>ocamlPackages.cohttp-mirage</li>
    <li>ocamlPackages.cohttp-top</li>
    <li>ocamlPackages.cohttp_async_websocket</li>
    <li>ocamlPackages.cohttp_static_handler</li>
    <li>ocamlPackages.conduit</li>
    <li>ocamlPackages.conduit-async</li>
    <li>ocamlPackages.conduit-lwt</li>
    <li>ocamlPackages.conduit-lwt-unix</li>
    <li>ocamlPackages.conduit-mirage</li>
    <li>ocamlPackages.containers</li>
    <li>ocamlPackages.containers-data</li>
    <li>ocamlPackages.content_security_policy</li>
    <li>ocamlPackages.cooltt</li>
    <li>ocamlPackages.core</li>
    <li>ocamlPackages.core_bench</li>
    <li>ocamlPackages.core_extended</li>
    <li>ocamlPackages.core_kernel</li>
    <li>ocamlPackages.core_unix</li>
    <li>ocamlPackages.cow</li>
    <li>ocamlPackages.cstruct-async</li>
    <li>ocamlPackages.cstruct-sexp</li>
    <li>ocamlPackages.csvfields</li>
    <li>ocamlPackages.ctypes_stubs_js</li>
    <li>ocamlPackages.curly</li>
    <li>ocamlPackages.dap</li>
    <li>ocamlPackages.data-encoding</li>
    <li>ocamlPackages.dbf</li>
    <li>ocamlPackages.decompress</li>
    <li>ocamlPackages.dedent</li>
    <li>ocamlPackages.delimited_parsing</li>
    <li>ocamlPackages.diffable</li>
    <li>ocamlPackages.digestif</li>
    <li>ocamlPackages.dns</li>
    <li>ocamlPackages.dns-certify</li>
    <li>ocamlPackages.dns-cli</li>
    <li>ocamlPackages.dns-client</li>
    <li>ocamlPackages.dns-client-lwt</li>
    <li>ocamlPackages.dns-client-mirage</li>
    <li>ocamlPackages.dns-mirage</li>
    <li>ocamlPackages.dns-resolver</li>
    <li>ocamlPackages.dns-server</li>
    <li>ocamlPackages.dns-stub</li>
    <li>ocamlPackages.dns-tsig</li>
    <li>ocamlPackages.dnssec</li>
    <li>ocamlPackages.domain-local-timeout</li>
    <li>ocamlPackages.dose3</li>
    <li>ocamlPackages.dune-action-plugin</li>
    <li>ocamlPackages.dune-build-info</li>
    <li>ocamlPackages.dune-rpc</li>
    <li>ocamlPackages.ecaml</li>
    <li>ocamlPackages.eliom</li>
    <li>ocamlPackages.elpi</li>
    <li>ocamlPackages.email_message</li>
    <li>ocamlPackages.emile</li>
    <li>ocamlPackages.encore</li>
    <li>ocamlPackages.env_config</li>
    <li>ocamlPackages.erm_xmpp</li>
    <li>ocamlPackages.ethernet</li>
    <li>ocamlPackages.expect_test_helpers_async</li>
    <li>ocamlPackages.expect_test_helpers_core</li>
    <li>ocamlPackages.ezjsonm</li>
    <li>ocamlPackages.faraday-async</li>
    <li>ocamlPackages.fieldslib</li>
    <li>ocamlPackages.file_path</li>
    <li>ocamlPackages.flex</li>
    <li>ocamlPackages.functoria</li>
    <li>ocamlPackages.fuzzy_match</li>
    <li>ocamlPackages.fzf</li>
    <li>ocamlPackages.gapi-ocaml</li>
    <li>ocamlPackages.gen_js_api</li>
    <li>ocamlPackages.genspio</li>
    <li>ocamlPackages.git</li>
    <li>ocamlPackages.git-mirage</li>
    <li>ocamlPackages.git-paf</li>
    <li>ocamlPackages.git-unix</li>
    <li>ocamlPackages.github</li>
    <li>ocamlPackages.github-jsoo</li>
    <li>ocamlPackages.github-unix</li>
    <li>ocamlPackages.google-drive-ocamlfuse</li>
    <li>ocamlPackages.graphql-cohttp</li>
    <li>ocamlPackages.graphql_ppx</li>
    <li>ocamlPackages.h2</li>
    <li>ocamlPackages.hack_parallel</li>
    <li>ocamlPackages.happy-eyeballs</li>
    <li>ocamlPackages.happy-eyeballs-lwt</li>
    <li>ocamlPackages.happy-eyeballs-mirage</li>
    <li>ocamlPackages.higher_kinded</li>
    <li>ocamlPackages.higlo</li>
    <li>ocamlPackages.hpack</li>
    <li>ocamlPackages.http-mirage-client</li>
    <li>ocamlPackages.httpaf</li>
    <li>ocamlPackages.httpaf-lwt-unix</li>
    <li>ocamlPackages.imagelib</li>
    <li>ocamlPackages.incr_dom</li>
    <li>ocamlPackages.incr_map</li>
    <li>ocamlPackages.incr_select</li>
    <li>ocamlPackages.incremental</li>
    <li>ocamlPackages.indentation_buffer</li>
    <li>ocamlPackages.index</li>
    <li>ocamlPackages.int_repr</li>
    <li>ocamlPackages.integers_stubs_js</li>
    <li>ocamlPackages.ipaddr</li>
    <li>ocamlPackages.ipaddr-cstruct</li>
    <li>ocamlPackages.ipaddr-sexp</li>
    <li>ocamlPackages.iri</li>
    <li>ocamlPackages.irmin</li>
    <li>ocamlPackages.irmin-chunk</li>
    <li>ocamlPackages.irmin-containers</li>
    <li>ocamlPackages.irmin-fs</li>
    <li>ocamlPackages.irmin-git</li>
    <li>ocamlPackages.irmin-graphql</li>
    <li>ocamlPackages.irmin-http</li>
    <li>ocamlPackages.irmin-mirage</li>
    <li>ocamlPackages.irmin-mirage-git</li>
    <li>ocamlPackages.irmin-mirage-graphql</li>
    <li>ocamlPackages.irmin-pack</li>
    <li>ocamlPackages.irmin-test</li>
    <li>ocamlPackages.irmin-tezos</li>
    <li>ocamlPackages.irmin-watcher</li>
    <li>ocamlPackages.iter</li>
    <li>ocamlPackages.jane-street-headers</li>
    <li>ocamlPackages.jane_rope</li>
    <li>ocamlPackages.janestreet_cpuid</li>
    <li>ocamlPackages.janestreet_csv</li>
    <li>ocamlPackages.jingoo</li>
    <li>ocamlPackages.js_of_ocaml</li>
    <li>ocamlPackages.js_of_ocaml-compiler</li>
    <li>ocamlPackages.js_of_ocaml-lwt</li>
    <li>ocamlPackages.js_of_ocaml-ppx</li>
    <li>ocamlPackages.js_of_ocaml-ppx_deriving_json</li>
    <li>ocamlPackages.js_of_ocaml-toplevel</li>
    <li>ocamlPackages.js_of_ocaml-tyxml</li>
    <li>ocamlPackages.js_of_ocaml_patches</li>
    <li>ocamlPackages.json-data-encoding</li>
    <li>ocamlPackages.json-data-encoding-bson</li>
    <li>ocamlPackages.jsonaf</li>
    <li>ocamlPackages.jst-config</li>
    <li>ocamlPackages.jwto</li>
    <li>ocamlPackages.lambda-term</li>
    <li>ocamlPackages.lambdapi</li>
    <li>ocamlPackages.lens</li>
    <li>ocamlPackages.letsencrypt</li>
    <li>ocamlPackages.letsencrypt-app</li>
    <li>ocamlPackages.letsencrypt-dns</li>
    <li>ocamlPackages.letsencrypt-mirage</li>
    <li>ocamlPackages.linol</li>
    <li>ocamlPackages.linol-lwt</li>
    <li>ocamlPackages.logs</li>
    <li>ocamlPackages.lru_cache</li>
    <li>ocamlPackages.lsp</li>
    <li>ocamlPackages.lustre-v6</li>
    <li>ocamlPackages.lwt_ppx</li>
    <li>ocamlPackages.macaddr</li>
    <li>ocamlPackages.macaddr-cstruct</li>
    <li>ocamlPackages.macaddr-sexp</li>
    <li>ocamlPackages.macaque</li>
    <li>ocamlPackages.man_in_the_middle_debugger</li>
    <li>ocamlPackages.mdx</li>
    <li>ocamlPackages.mdx.bin</li>
    <li>ocamlPackages.mdx.lib</li>
    <li>ocamlPackages.metrics-lwt</li>
    <li>ocamlPackages.metrics-rusage</li>
    <li>ocamlPackages.metrics-unix</li>
    <li>ocamlPackages.mimic</li>
    <li>ocamlPackages.mimic-happy-eyeballs</li>
    <li>ocamlPackages.mirage</li>
    <li>ocamlPackages.mirage-block-combinators</li>
    <li>ocamlPackages.mirage-block-unix</li>
    <li>ocamlPackages.mirage-bootvar-xen</li>
    <li>ocamlPackages.mirage-channel</li>
    <li>ocamlPackages.mirage-crypto-ec</li>
    <li>ocamlPackages.mirage-crypto-pk</li>
    <li>ocamlPackages.mirage-crypto-rng</li>
    <li>ocamlPackages.mirage-crypto-rng-async</li>
    <li>ocamlPackages.mirage-crypto-rng-lwt</li>
    <li>ocamlPackages.mirage-crypto-rng-mirage</li>
    <li>ocamlPackages.mirage-flow-combinators</li>
    <li>ocamlPackages.mirage-flow-unix</li>
    <li>ocamlPackages.mirage-logs</li>
    <li>ocamlPackages.mirage-nat</li>
    <li>ocamlPackages.mirage-net</li>
    <li>ocamlPackages.mirage-net-xen</li>
    <li>ocamlPackages.mirage-profile</li>
    <li>ocamlPackages.mirage-protocols</li>
    <li>ocamlPackages.mirage-runtime</li>
    <li>ocamlPackages.mirage-stack</li>
    <li>ocamlPackages.mirage-unix</li>
    <li>ocamlPackages.mirage-vnetif</li>
    <li>ocamlPackages.mirage-xen</li>
    <li>ocamlPackages.mirage.dev</li>
    <li>ocamlPackages.mldoc</li>
    <li>ocamlPackages.morbig</li>
    <li>ocamlPackages.mrmime</li>
    <li>ocamlPackages.multipart-form-data</li>
    <li>ocamlPackages.mustache</li>
    <li>ocamlPackages.n_ary</li>
    <li>ocamlPackages.netchannel</li>
    <li>ocamlPackages.numeric_string</li>
    <li>ocamlPackages.ocaml-embed-file</li>
    <li>ocamlPackages.ocaml-lsp</li>
    <li>ocamlPackages.ocaml-monadic</li>
    <li>ocamlPackages.ocaml-print-intf</li>
    <li>ocamlPackages.ocaml-r</li>
    <li>ocamlPackages.ocaml-vdom</li>
    <li>ocamlPackages.ocaml_intrinsics</li>
    <li>ocamlPackages.ocamlformat (ocamlPackages.ocamlformat_0_26_0)</li>
    <li>ocamlPackages.ocamlformat-lib</li>
    <li>ocamlPackages.ocamlformat-rpc-lib</li>
    <li>ocamlPackages.ocamlformat_0_19_0</li>
    <li>ocamlPackages.ocamlformat_0_20_0</li>
    <li>ocamlPackages.ocamlformat_0_20_1</li>
    <li>ocamlPackages.ocamlformat_0_21_0</li>
    <li>ocamlPackages.ocamlformat_0_22_4</li>
    <li>ocamlPackages.ocamlformat_0_23_0</li>
    <li>ocamlPackages.ocamlformat_0_24_1</li>
    <li>ocamlPackages.ocamlformat_0_25_1</li>
    <li>ocamlPackages.ocf_ppx</li>
    <li>ocamlPackages.ocsigen-ppx-rpc</li>
    <li>ocamlPackages.ocsigen-start</li>
    <li>ocamlPackages.ocsigen-toolkit</li>
    <li>ocamlPackages.ocsigen_server</li>
    <li>ocamlPackages.ocsipersist</li>
    <li>ocamlPackages.ocsipersist-lib</li>
    <li>ocamlPackages.ocsipersist-pgsql</li>
    <li>ocamlPackages.ocsipersist-sqlite</li>
    <li>ocamlPackages.ocurl</li>
    <li>ocamlPackages.of_json</li>
    <li>ocamlPackages.ojs</li>
    <li>ocamlPackages.opium</li>
    <li>ocamlPackages.ordinal_abbreviation</li>
    <li>ocamlPackages.oseq</li>
    <li>ocamlPackages.otr</li>
    <li>ocamlPackages.owl</li>
    <li>ocamlPackages.paf</li>
    <li>ocamlPackages.paf-cohttp</li>
    <li>ocamlPackages.parsexp</li>
    <li>ocamlPackages.patdiff</li>
    <li>ocamlPackages.patience_diff</li>
    <li>ocamlPackages.pcap-format</li>
    <li>ocamlPackages.pgocaml</li>
    <li>ocamlPackages.pgocaml_ppx</li>
    <li>ocamlPackages.piaf</li>
    <li>ocamlPackages.piqi</li>
    <li>ocamlPackages.piqi-ocaml</li>
    <li>ocamlPackages.plotkicadsch</li>
    <li>ocamlPackages.polling_state_rpc</li>
    <li>ocamlPackages.posixat</li>
    <li>ocamlPackages.pp</li>
    <li>ocamlPackages.ppx_accessor</li>
    <li>ocamlPackages.ppx_assert</li>
    <li>ocamlPackages.ppx_base</li>
    <li>ocamlPackages.ppx_bench</li>
    <li>ocamlPackages.ppx_bin_prot</li>
    <li>ocamlPackages.ppx_bitstring</li>
    <li>ocamlPackages.ppx_blob</li>
    <li>ocamlPackages.ppx_cold</li>
    <li>ocamlPackages.ppx_compare</li>
    <li>ocamlPackages.ppx_css</li>
    <li>ocamlPackages.ppx_cstruct</li>
    <li>ocamlPackages.ppx_cstubs</li>
    <li>ocamlPackages.ppx_custom_printf</li>
    <li>ocamlPackages.ppx_demo</li>
    <li>ocamlPackages.ppx_derive_at_runtime</li>
    <li>ocamlPackages.ppx_deriving</li>
    <li>ocamlPackages.ppx_deriving_cmdliner</li>
    <li>ocamlPackages.ppx_deriving_protobuf</li>
    <li>ocamlPackages.ppx_deriving_qcheck</li>
    <li>ocamlPackages.ppx_deriving_rpc</li>
    <li>ocamlPackages.ppx_deriving_yaml</li>
    <li>ocamlPackages.ppx_deriving_yojson</li>
    <li>ocamlPackages.ppx_disable_unused_warnings</li>
    <li>ocamlPackages.ppx_enumerate</li>
    <li>ocamlPackages.ppx_expect</li>
    <li>ocamlPackages.ppx_fields_conv</li>
    <li>ocamlPackages.ppx_fixed_literal</li>
    <li>ocamlPackages.ppx_gen_rec</li>
    <li>ocamlPackages.ppx_globalize</li>
    <li>ocamlPackages.ppx_hash</li>
    <li>ocamlPackages.ppx_here</li>
    <li>ocamlPackages.ppx_ignore_instrumentation</li>
    <li>ocamlPackages.ppx_import</li>
    <li>ocamlPackages.ppx_inline_test</li>
    <li>ocamlPackages.ppx_irmin</li>
    <li>ocamlPackages.ppx_jane</li>
    <li>ocamlPackages.ppx_js_style</li>
    <li>ocamlPackages.ppx_jsonaf_conv</li>
    <li>ocamlPackages.ppx_let</li>
    <li>ocamlPackages.ppx_log</li>
    <li>ocamlPackages.ppx_module_timer</li>
    <li>ocamlPackages.ppx_monad</li>
    <li>ocamlPackages.ppx_optcomp</li>
    <li>ocamlPackages.ppx_optional</li>
    <li>ocamlPackages.ppx_pattern_bind</li>
    <li>ocamlPackages.ppx_pipebang</li>
    <li>ocamlPackages.ppx_python</li>
    <li>ocamlPackages.ppx_repr</li>
    <li>ocamlPackages.ppx_sexp_conv</li>
    <li>ocamlPackages.ppx_sexp_message</li>
    <li>ocamlPackages.ppx_sexp_value</li>
    <li>ocamlPackages.ppx_show</li>
    <li>ocamlPackages.ppx_stable</li>
    <li>ocamlPackages.ppx_stable_witness</li>
    <li>ocamlPackages.ppx_string</li>
    <li>ocamlPackages.ppx_tydi</li>
    <li>ocamlPackages.ppx_typed_fields</li>
    <li>ocamlPackages.ppx_typerep_conv</li>
    <li>ocamlPackages.ppx_variants_conv</li>
    <li>ocamlPackages.ppx_yojson_conv</li>
    <li>ocamlPackages.ppxlib</li>
    <li>ocamlPackages.prettym</li>
    <li>ocamlPackages.printbox</li>
    <li>ocamlPackages.printbox-text</li>
    <li>ocamlPackages.profunctor</li>
    <li>ocamlPackages.progress</li>
    <li>ocamlPackages.promise_jsoo</li>
    <li>ocamlPackages.protocol_version_header</li>
    <li>ocamlPackages.pyml</li>
    <li>ocamlPackages.re2</li>
    <li>ocamlPackages.re2_stable</li>
    <li>ocamlPackages.reason</li>
    <li>ocamlPackages.reason-native.console</li>
    <li>ocamlPackages.reason-native.dir</li>
    <li>ocamlPackages.reason-native.file-context-printer</li>
    <li>ocamlPackages.reason-native.fp</li>
    <li>ocamlPackages.reason-native.pastel</li>
    <li>ocamlPackages.reason-native.pastel-console</li>
    <li>ocamlPackages.reason-native.qcheck-rely</li>
    <li>ocamlPackages.reason-native.refmterr</li>
    <li>ocamlPackages.reason-native.rely</li>
    <li>ocamlPackages.reason-native.rely-junit-reporter</li>
    <li>ocamlPackages.rebez</li>
    <li>ocamlPackages.record_builder</li>
    <li>ocamlPackages.redis-async</li>
    <li>ocamlPackages.regex_parser_intf</li>
    <li>ocamlPackages.reperf</li>
    <li>ocamlPackages.repr</li>
    <li>ocamlPackages.resource_cache</li>
    <li>ocamlPackages.resto</li>
    <li>ocamlPackages.resto-acl</li>
    <li>ocamlPackages.resto-cohttp</li>
    <li>ocamlPackages.resto-cohttp-client</li>
    <li>ocamlPackages.resto-cohttp-self-serving-client</li>
    <li>ocamlPackages.resto-cohttp-server</li>
    <li>ocamlPackages.resto-directory</li>
    <li>ocamlPackages.resto-json</li>
    <li>ocamlPackages.rock</li>
    <li>ocamlPackages.rosetta</li>
    <li>ocamlPackages.rpclib</li>
    <li>ocamlPackages.rpclib-lwt</li>
    <li>ocamlPackages.sail</li>
    <li>ocamlPackages.secp256k1</li>
    <li>ocamlPackages.sedlex</li>
    <li>ocamlPackages.semantic_version</li>
    <li>ocamlPackages.sexp</li>
    <li>ocamlPackages.sexp_diff</li>
    <li>ocamlPackages.sexp_grammar</li>
    <li>ocamlPackages.sexp_macro</li>
    <li>ocamlPackages.sexp_pretty</li>
    <li>ocamlPackages.sexp_select</li>
    <li>ocamlPackages.sexplib</li>
    <li>ocamlPackages.sexplib0</li>
    <li>ocamlPackages.shared-memory-ring</li>
    <li>ocamlPackages.shared-memory-ring-lwt</li>
    <li>ocamlPackages.shell</li>
    <li>ocamlPackages.shexp</li>
    <li>ocamlPackages.spawn</li>
    <li>ocamlPackages.splay_tree</li>
    <li>ocamlPackages.splittable_random</li>
    <li>ocamlPackages.stdio</li>
    <li>ocamlPackages.stored_reversed</li>
    <li>ocamlPackages.streamable</li>
    <li>ocamlPackages.tar</li>
    <li>ocamlPackages.tar-unix</li>
    <li>ocamlPackages.tcpip</li>
    <li>ocamlPackages.tdigest</li>
    <li>ocamlPackages.telegraml</li>
    <li>ocamlPackages.textutils</li>
    <li>ocamlPackages.textutils_kernel</li>
    <li>ocamlPackages.tezos-base58</li>
    <li>ocamlPackages.thread-table</li>
    <li>ocamlPackages.tilde_f</li>
    <li>ocamlPackages.time_now</li>
    <li>ocamlPackages.timezone</li>
    <li>ocamlPackages.tls</li>
    <li>ocamlPackages.tls-lwt</li>
    <li>ocamlPackages.tls-mirage</li>
    <li>ocamlPackages.topological_sort</li>
    <li>ocamlPackages.torch</li>
    <li>ocamlPackages.tsort</li>
    <li>ocamlPackages.tuntap</li>
    <li>ocamlPackages.typerep</li>
    <li>ocamlPackages.tyxml-lwd</li>
    <li>ocamlPackages.unstrctrd</li>
    <li>ocamlPackages.uri</li>
    <li>ocamlPackages.uri-sexp</li>
    <li>ocamlPackages.uring</li>
    <li>ocamlPackages.utop</li>
    <li>ocamlPackages.uuuu</li>
    <li>ocamlPackages.variantslib</li>
    <li>ocamlPackages.vcaml</li>
    <li>ocamlPackages.vchan</li>
    <li>ocamlPackages.vg</li>
    <li>ocamlPackages.virtual_dom</li>
    <li>ocamlPackages.visitors</li>
    <li>ocamlPackages.wayland</li>
    <li>ocamlPackages.webbrowser</li>
    <li>ocamlPackages.webmachine</li>
    <li>ocamlPackages.x509</li>
    <li>ocamlPackages.xenstore</li>
    <li>ocamlPackages.xenstore-tool</li>
    <li>ocamlPackages.xenstore_transport</li>
    <li>ocamlPackages.xtmpl</li>
    <li>ocamlPackages.xtmpl_ppx</li>
    <li>ocamlPackages.yaml</li>
    <li>ocamlPackages.yaml-sexp</li>
    <li>ocamlPackages.zarith_stubs_js</li>
    <li>ocamlPackages.zstandard</li>
    <li>python310Packages.bap</li>
    <li>python310Packages.bap.dist</li>
    <li>python311Packages.bap</li>
    <li>python311Packages.bap.dist</li>
    <li>sail-riscv-rv32</li>
    <li>sail-riscv-rv64</li>
    <li>satysfi</li>
    <li>soupault</li>
    <li>wayland-proxy-virtwl</li>
    <li>why3</li>
  </ul>
</details>

vbgl commented on pkgs/top-level/ocaml-packages.nix:
```diff
@@ -1836,6 +1871,7 @@ in let inherit (pkgs) callPackage; in rec
   ocamlPackages_latest = ocamlPackages_5_0;
 
   ocamlPackages = ocamlPackages_4_14;
+  ocamlPackages_4_14_janeStreet_0_15 = ocamlPackages_4_14.overrideScope' (self: super: super // super.janeStreet_0_15);
```
Is this really needed? It does not look right.

dimitrijer replied:
I needed this in order to build packages outside of `ocamlPackages` that depend on janestreet v0.15 libraries AND other non-JS OCaml packages, like `stanc`. With `overrideScope'`, any non-JS OCaml modules that depend on JS libraries will pull in v0.15 libraries instead of the default v0.16 libraries. That way, external packages can use both JS and non-JS libraries without accidentally pulling in both v0.15 and v0.16 due to transitive build dependencies. Simply plug in `ocamlPackages_4_14_janeStreet_0_15` as `ocamlPackages` in stanc derivation.

The idea is to override attributes lifted one level up from `janeStreet` (v0.16) by first call to `overrideScope'`, with attributes from `janeStreet_0_15` (v0.15).

Another way of dealing with this is carefully and incrementally crafting transitive dependencies with `.override` to use v0.15, like I had to do for e.g. `bap`. But this seems quite messy.

It's nasty, I agree. I reckon we can get rid of this as soon as v0.16 libraries are supported upstream.
Any cleaner solutions to this are very welcome!

vbgl replied:
Alright. If we agree on the fact that this is a temporary ugly hack, then that’s fine. Let’s hope `stanc` and `ligo` get updated.

vbgl commented:
@GrahamcOfBorg build ligo

vbgl commented:
Evaluation seems broken.

dimitrijer commented:
> 6 packages failed to build:
> 
>     * ocamlPackages.biocaml
> 
>     * ocamlPackages.bistro
> 
>     * ocamlPackages.cfstream
> 
>     * ocamlPackages.magic-trace
> 
>     * ocamlPackages.phylogenetics
> 
>     * ocamlPackages.tls-async

Working on getting these to compile. `bistro` and `cfstream` were fairly easy to fix. `biocaml` is proving to be a lot of work, I'll probably just move it under `janeStreet_0_15` just like I did for `bap`.

dimitrijer commented:
Evaluation now builds. All packages affected by the PR now build on x86_64-linux -- tested with `nix-shell -p nixpkgs-review --run "nixpkgs-review pr 247022"`.

Due to changes to `tls-async`, this PR now affects more packages: 31 packages added, 502 packages updated.

 `ocamlPackages.uring` hangs in test phase, but this was because I was running the review on an older kernel (`3.10.0-1160.95.1.el7.x86_64`). It correctly builds on a newer kernel (`6.4.12-arch1-1`).

Kindly review the changes.

dimitrijer commented:
Output of `nix-shell -p nixpkgs-review --run "nixpkgs-review pr 247022"` (`ocamlPackages.uring` builds successfully on newer kernels):

```text
1 package marked as broken and skipped:
ocamlPackages.pythonlib

1 package failed to build:
ocamlPackages.uring

535 packages built:
acgtk advi alt-ergo beluga coccinelle comby coqPackages.coq-elpi coqPackages.coq-lsp coqPackages.hierarchy-builder coqPackages.mathcomp-algebra-tactics coqPackages.mathcomp-analysis coqPackages.mathcomp-classical coqPackages.mathcomp-infotheo coqPackages.serapi coqPackages.trakt cvehound cvehound.dist dune-release easy
crypt eff flitter flow framac fstar glsurf hol_light jackline ledit libbap liquidsoap merge-fmt ocamlPackages.abstract_algebra ocamlPackages.accessor ocamlPackages.accessor_async ocamlPackages.accessor_base ocamlPackages.accessor_core ocamlPackages.alcotest-lwt ocamlPackages.alcotest-mirage ocamlPackages.angstrom ocaml
Packages.angstrom-async ocamlPackages.angstrom-lwt-unix ocamlPackages.angstrom-unix ocamlPackages.arp ocamlPackages.async ocamlPackages.async_durable ocamlPackages.async_extra ocamlPackages.async_find ocamlPackages.async_inotify ocamlPackages.async_interactive ocamlPackages.async_js ocamlPackages.async_kernel ocamlPack
ages.async_rpc_kernel ocamlPackages.async_rpc_websocket ocamlPackages.async_sendfile ocamlPackages.async_shell ocamlPackages.async_smtp ocamlPackages.async_ssl ocamlPackages.async_unix ocamlPackages.async_websocket ocamlPackages.awa ocamlPackages.awa-mirage ocamlPackages.babel ocamlPackages.bap ocamlPackages.base ocaml
Packages.base64 ocamlPackages.base_bigstring ocamlPackages.base_quickcheck ocamlPackages.base_trie ocamlPackages.bidirectional_map ocamlPackages.bignum ocamlPackages.bin_prot ocamlPackages.biocaml ocamlPackages.bisect_ppx ocamlPackages.bistro ocamlPackages.bls12-381 ocamlPackages.bls12-381-signature ocamlPackages.bonsa
i ocamlPackages.bos ocamlPackages.brisk-reconciler ocamlPackages.brr ocamlPackages.ca-certs ocamlPackages.ca-certs-nss ocamlPackages.camlimages ocamlPackages.camlp5 ocamlPackages.caqti ocamlPackages.caqti-async ocamlPackages.caqti-driver-mariadb ocamlPackages.caqti-driver-postgresql ocamlPackages.caqti-driver-sqlite3 o
camlPackages.caqti-dynload ocamlPackages.caqti-lwt ocamlPackages.caqti-type-calendar ocamlPackages.carton ocamlPackages.carton-git ocamlPackages.carton-lwt ocamlPackages.cfstream ocamlPackages.checkseum ocamlPackages.cinaps ocamlPackages.class_group_vdf ocamlPackages.cohttp ocamlPackages.cohttp-async ocamlPackages.coht
tp-lwt ocamlPackages.cohttp-lwt-jsoo ocamlPackages.cohttp-lwt-unix ocamlPackages.cohttp-mirage ocamlPackages.cohttp-top ocamlPackages.cohttp_async_websocket ocamlPackages.cohttp_static_handler ocamlPackages.conduit ocamlPackages.conduit-async ocamlPackages.conduit-lwt ocamlPackages.conduit-lwt-unix ocamlPackages.condui
t-mirage ocamlPackages.containers ocamlPackages.containers-data ocamlPackages.content_security_policy ocamlPackages.cooltt ocamlPackages.core ocamlPackages.core_bench ocamlPackages.core_extended ocamlPackages.core_kernel ocamlPackages.core_unix ocamlPackages.cow ocamlPackages.cstruct-async ocamlPackages.cstruct-sexp oc
amlPackages.csvfields ocamlPackages.ctypes_stubs_js ocamlPackages.curly ocamlPackages.dap ocamlPackages.data-encoding ocamlPackages.dbf ocamlPackages.decompress ocamlPackages.dedent ocamlPackages.delimited_parsing ocamlPackages.diffable ocamlPackages.digestif ocamlPackages.dns ocamlPackages.dns-certify ocamlPackages.dn
s-cli ocamlPackages.dns-client ocamlPackages.dns-client-lwt ocamlPackages.dns-client-mirage ocamlPackages.dns-mirage ocamlPackages.dns-resolver ocamlPackages.dns-server ocamlPackages.dns-stub ocamlPackages.dns-tsig ocamlPackages.dnssec ocamlPackages.domain-local-timeout ocamlPackages.dose3 ocamlPackages.dune-action-plu
gin ocamlPackages.dune-build-info ocamlPackages.dune-rpc ocamlPackages.ecaml ocamlPackages.eliom ocamlPackages.elpi ocamlPackages.email_message ocamlPackages.emile ocamlPackages.encore ocamlPackages.env_config ocamlPackages.erm_xmpp ocamlPackages.ethernet ocamlPackages.expect_test_helpers_async ocamlPackages.expect_tes
t_helpers_core ocamlPackages.ezjsonm ocamlPackages.faraday-async ocamlPackages.fieldslib ocamlPackages.file_path ocamlPackages.flex ocamlPackages.functoria ocamlPackages.fuzzy_match ocamlPackages.fzf ocamlPackages.gapi-ocaml ocamlPackages.gen_js_api ocamlPackages.genspio ocamlPackages.git ocamlPackages.git-mirage ocaml
Packages.git-paf ocamlPackages.git-unix ocamlPackages.github ocamlPackages.github-jsoo ocamlPackages.github-unix ocamlPackages.google-drive-ocamlfuse ocamlPackages.graphql-cohttp ocamlPackages.graphql_ppx ocamlPackages.h2 ocamlPackages.hack_parallel ocamlPackages.happy-eyeballs ocamlPackages.happy-eyeballs-lwt ocamlPac
kages.happy-eyeballs-mirage ocamlPackages.higher_kinded ocamlPackages.higlo ocamlPackages.hpack ocamlPackages.http-mirage-client ocamlPackages.httpaf ocamlPackages.httpaf-lwt-unix ocamlPackages.imagelib ocamlPackages.incr_dom ocamlPackages.incr_map ocamlPackages.incr_select ocamlPackages.incremental ocamlPackages.inden
tation_buffer ocamlPackages.index ocamlPackages.int_repr ocamlPackages.integers_stubs_js ocamlPackages.ipaddr ocamlPackages.ipaddr-cstruct ocamlPackages.ipaddr-sexp ocamlPackages.iri ocamlPackages.irmin ocamlPackages.irmin-chunk ocamlPackages.irmin-containers ocamlPackages.irmin-fs ocamlPackages.irmin-git ocamlPackages
.irmin-graphql ocamlPackages.irmin-http ocamlPackages.irmin-mirage ocamlPackages.irmin-mirage-git ocamlPackages.irmin-mirage-graphql ocamlPackages.irmin-pack ocamlPackages.irmin-test ocamlPackages.irmin-tezos ocamlPackages.irmin-watcher ocamlPackages.iter ocamlPackages.jane-street-headers ocamlPackages.jane_rope ocamlP
ackages.janestreet_cpuid ocamlPackages.janestreet_csv ocamlPackages.jingoo ocamlPackages.js_of_ocaml ocamlPackages.js_of_ocaml-compiler ocamlPackages.js_of_ocaml-lwt ocamlPackages.js_of_ocaml-ppx ocamlPackages.js_of_ocaml-ppx_deriving_json ocamlPackages.js_of_ocaml-toplevel ocamlPackages.js_of_ocaml-tyxml ocamlPackages
.js_of_ocaml_patches ocamlPackages.json-data-encoding ocamlPackages.json-data-encoding-bson ocamlPackages.jsonaf ocamlPackages.jst-config ocamlPackages.jwto ocamlPackages.lambda-term ocamlPackages.lambdapi ocamlPackages.lens ocamlPackages.letsencrypt ocamlPackages.letsencrypt-app ocamlPackages.letsencrypt-dns ocamlPack
ages.letsencrypt-mirage ocamlPackages.linol ocamlPackages.linol-lwt ocamlPackages.logs ocamlPackages.lru_cache ocamlPackages.lsp ocamlPackages.lustre-v6 ocamlPackages.lwt_ppx ocamlPackages.macaddr ocamlPackages.macaddr-cstruct ocamlPackages.macaddr-sexp ocamlPackages.macaque ocamlPackages.magic-trace ocamlPackages.man_
in_the_middle_debugger ocamlPackages.mdx ocamlPackages.mdx.bin ocamlPackages.mdx.lib ocamlPackages.metrics-lwt ocamlPackages.metrics-rusage ocamlPackages.metrics-unix ocamlPackages.mimic ocamlPackages.mimic-happy-eyeballs ocamlPackages.mirage ocamlPackages.mirage-block-combinators ocamlPackages.mirage-block-unix ocamlP
ackages.mirage-bootvar-xen ocamlPackages.mirage-channel ocamlPackages.mirage-crypto-ec ocamlPackages.mirage-crypto-pk ocamlPackages.mirage-crypto-rng ocamlPackages.mirage-crypto-rng-async ocamlPackages.mirage-crypto-rng-lwt ocamlPackages.mirage-crypto-rng-mirage ocamlPackages.mirage-flow-combinators ocamlPackages.mirag
e-flow-unix ocamlPackages.mirage-logs ocamlPackages.mirage-nat ocamlPackages.mirage-net ocamlPackages.mirage-net-xen ocamlPackages.mirage-profile ocamlPackages.mirage-protocols ocamlPackages.mirage-runtime ocamlPackages.mirage-stack ocamlPackages.mirage-unix ocamlPackages.mirage-vnetif ocamlPackages.mirage-xen ocamlPac
kages.mirage.dev ocamlPackages.mldoc ocamlPackages.morbig ocamlPackages.mrmime ocamlPackages.multipart-form-data ocamlPackages.mustache ocamlPackages.n_ary ocamlPackages.netchannel ocamlPackages.numeric_string ocamlPackages.ocaml-embed-file ocamlPackages.ocaml-lsp ocamlPackages.ocaml-monadic ocamlPackages.ocaml-print-i
ntf ocamlPackages.ocaml-protoc-plugin ocamlPackages.ocaml-r ocamlPackages.ocaml-vdom ocamlPackages.ocaml_intrinsics ocamlPackages.ocamlformat ocamlPackages.ocamlformat-lib ocamlPackages.ocamlformat-rpc-lib ocamlPackages.ocamlformat_0_19_0 ocamlPackages.ocamlformat_0_20_0 ocamlPackages.ocamlformat_0_20_1 ocamlPackages.o
camlformat_0_21_0 ocamlPackages.ocamlformat_0_22_4 ocamlPackages.ocamlformat_0_23_0 ocamlPackages.ocamlformat_0_24_1 ocamlPackages.ocamlformat_0_25_1 ocamlPackages.ocf_ppx ocamlPackages.ocsigen-ppx-rpc ocamlPackages.ocsigen-start ocamlPackages.ocsigen-toolkit ocamlPackages.ocsigen_server ocamlPackages.ocsipersist ocaml
Packages.ocsipersist-lib ocamlPackages.ocsipersist-pgsql ocamlPackages.ocsipersist-sqlite ocamlPackages.ocurl ocamlPackages.of_json ocamlPackages.ojs ocamlPackages.opium ocamlPackages.ordinal_abbreviation ocamlPackages.oseq ocamlPackages.otr ocamlPackages.owl ocamlPackages.paf ocamlPackages.paf-cohttp ocamlPackages.par
sexp ocamlPackages.patdiff ocamlPackages.patience_diff ocamlPackages.pcap-format ocamlPackages.pgocaml ocamlPackages.pgocaml_ppx ocamlPackages.phylogenetics ocamlPackages.piaf ocamlPackages.piqi ocamlPackages.piqi-ocaml ocamlPackages.plotkicadsch ocamlPackages.polling_state_rpc ocamlPackages.posixat ocamlPackages.pp oc
amlPackages.ppx_accessor ocamlPackages.ppx_assert ocamlPackages.ppx_base ocamlPackages.ppx_bench ocamlPackages.ppx_bin_prot ocamlPackages.ppx_bitstring ocamlPackages.ppx_blob ocamlPackages.ppx_cold ocamlPackages.ppx_compare ocamlPackages.ppx_css ocamlPackages.ppx_cstruct ocamlPackages.ppx_cstubs ocamlPackages.ppx_custo
m_printf ocamlPackages.ppx_demo ocamlPackages.ppx_derive_at_runtime ocamlPackages.ppx_deriving ocamlPackages.ppx_deriving_cmdliner ocamlPackages.ppx_deriving_protobuf ocamlPackages.ppx_deriving_qcheck ocamlPackages.ppx_deriving_rpc ocamlPackages.ppx_deriving_yaml ocamlPackages.ppx_deriving_yojson ocamlPackages.ppx_disa
ble_unused_warnings ocamlPackages.ppx_enumerate ocamlPackages.ppx_expect ocamlPackages.ppx_fields_conv ocamlPackages.ppx_fixed_literal ocamlPackages.ppx_gen_rec ocamlPackages.ppx_globalize ocamlPackages.ppx_hash ocamlPackages.ppx_here ocamlPackages.ppx_ignore_instrumentation ocamlPackages.ppx_import ocamlPackages.ppx_i
nline_test ocamlPackages.ppx_irmin ocamlPackages.ppx_jane ocamlPackages.ppx_js_style ocamlPackages.ppx_jsonaf_conv ocamlPackages.ppx_let ocamlPackages.ppx_log ocamlPackages.ppx_module_timer ocamlPackages.ppx_monad ocamlPackages.ppx_optcomp ocamlPackages.ppx_optional ocamlPackages.ppx_pattern_bind ocamlPackages.ppx_pipe
bang ocamlPackages.ppx_python ocamlPackages.ppx_repr ocamlPackages.ppx_sexp_conv ocamlPackages.ppx_sexp_message ocamlPackages.ppx_sexp_value ocamlPackages.ppx_show ocamlPackages.ppx_stable ocamlPackages.ppx_stable_witness ocamlPackages.ppx_string ocamlPackages.ppx_tydi ocamlPackages.ppx_typed_fields ocamlPackages.ppx_t
yperep_conv ocamlPackages.ppx_variants_conv ocamlPackages.ppx_yojson_conv ocamlPackages.ppxlib ocamlPackages.prettym ocamlPackages.printbox ocamlPackages.printbox-text ocamlPackages.profunctor ocamlPackages.progress ocamlPackages.promise_jsoo ocamlPackages.protocol_version_header ocamlPackages.pyml ocamlPackages.re2 oc
amlPackages.re2_stable ocamlPackages.reason ocamlPackages.reason-native.console ocamlPackages.reason-native.dir ocamlPackages.reason-native.file-context-printer ocamlPackages.reason-native.fp ocamlPackages.reason-native.pastel ocamlPackages.reason-native.pastel-console ocamlPackages.reason-native.qcheck-rely ocamlPacka
ges.reason-native.refmterr ocamlPackages.reason-native.rely ocamlPackages.reason-native.rely-junit-reporter ocamlPackages.rebez ocamlPackages.record_builder ocamlPackages.redis-async ocamlPackages.regex_parser_intf ocamlPackages.reperf ocamlPackages.repr ocamlPackages.resource_cache ocamlPackages.resto ocamlPackages.re
sto-acl ocamlPackages.resto-cohttp ocamlPackages.resto-cohttp-client ocamlPackages.resto-cohttp-self-serving-client ocamlPackages.resto-cohttp-server ocamlPackages.resto-directory ocamlPackages.resto-json ocamlPackages.rock ocamlPackages.rosetta ocamlPackages.rpclib ocamlPackages.rpclib-lwt ocamlPackages.sail ocamlPack
ages.secp256k1 ocamlPackages.sedlex ocamlPackages.sel ocamlPackages.semantic_version ocamlPackages.sexp ocamlPackages.sexp_diff ocamlPackages.sexp_grammar ocamlPackages.sexp_macro ocamlPackages.sexp_pretty ocamlPackages.sexp_select ocamlPackages.sexplib ocamlPackages.sexplib0 ocamlPackages.shared-memory-ring ocamlPacka
ges.shared-memory-ring-lwt ocamlPackages.shell ocamlPackages.shexp ocamlPackages.spawn ocamlPackages.splay_tree ocamlPackages.splittable_random ocamlPackages.stdio ocamlPackages.stored_reversed ocamlPackages.streamable ocamlPackages.tar ocamlPackages.tar-unix ocamlPackages.tcpip ocamlPackages.tdigest ocamlPackages.tele
graml ocamlPackages.textutils ocamlPackages.textutils_kernel ocamlPackages.tezos-base58 ocamlPackages.thread-table ocamlPackages.tilde_f ocamlPackages.time_now ocamlPackages.timezone ocamlPackages.tls ocamlPackages.tls-async ocamlPackages.tls-lwt ocamlPackages.tls-mirage ocamlPackages.topological_sort ocamlPackages.tor
ch ocamlPackages.tsort ocamlPackages.tuntap ocamlPackages.typerep ocamlPackages.tyxml-lwd ocamlPackages.unstrctrd ocamlPackages.uri ocamlPackages.uri-sexp ocamlPackages.utop ocamlPackages.uuuu ocamlPackages.variantslib ocamlPackages.vcaml ocamlPackages.vchan ocamlPackages.vg ocamlPackages.virtual_dom ocamlPackages.visi
tors ocamlPackages.wayland ocamlPackages.webbrowser ocamlPackages.webmachine ocamlPackages.x509 ocamlPackages.xenstore ocamlPackages.xenstore-tool ocamlPackages.xenstore_transport ocamlPackages.xtmpl ocamlPackages.xtmpl_ppx ocamlPackages.yaml ocamlPackages.yaml-sexp ocamlPackages.zarith_stubs_js ocamlPackages.zstandard
 python310Packages.bap python310Packages.bap.dist python311Packages.bap python311Packages.bap.dist sail-riscv-rv32 sail-riscv-rv64 satysfi soupault wayland-proxy-virtwl why3
```

vbgl reviewed (approved):
Would you like to clean the history? Otherwise, all commits may be squashed into a single one.

dimitrijer commented:
Yep, that's a good idea. I'll do it tonight when I'm back home.

dimitrijer committed:
ocamlPackages.janeStreet: 0.15 -> 0.16

vbgl commented on pkgs/development/ocaml-modules/tls/async.nix:
```diff
@@ -8,6 +8,11 @@ buildDunePackage rec {
   minimalOCamlVersion = "4.11";
   duneVersion = "3";
```
```suggestion
  minimalOCamlVersion = "4.13";
```

vbgl committed:
Update pkgs/development/ocaml-modules/tls/async.nix

anmonteiro commented:
Packages added in #230267 but not here:

```

async_udp
bigdecimal
core_profiler
hex_encode
hg_lib
incr_dom_interactive
incr_dom_partial_render
incr_dom_sexp_form
krb
line-up-words
mlt_parser
netsnmp
notty_async
ocaml-probes
pam
parsexp_io
postgres_async
ppx_conv_func
ppx_csv_conv
ppx_xml_conv
pythonlib
re_parser
rpc_parallel
sequencer_table
sexp_string_quickcheck
string_dict
toplevel_backend
toplevel_expect_test
tracing
username_kernel

```

vbgl commented:
Three of those missing packages have just been added in #252614.
